src/widgets/virtual_list.rs

```diff
@@ -0,0 +1,309 @@
+        ) -> Size<UPx> {
+        let item_height = self.item_height.get_tracking_invalidate(context);
+        let item_height_upx = item_height.into_upx(context.gfx.scale());
+        let item_count = self.item_count.get_tracking_invalidate(context);
+        let content_height = item_height * item_count as i32;
+        let content_height = content_height.into_upx(context.gfx.scale());
+        let width = self.width.get_tracking_invalidate(context);
+        let width = width.into_upx(context.gfx.scale());
+
+        let new_control_size = Size::new(
+            width,
+            constrain_child(available_space.height, content_height),
+        );
+
+        let vertical = self
+            .vertical_scroll
+            .make_if_needed()
+            .mounted(&mut context.as_event_context());
+        let scrollbar_layout = context.for_other(&vertical).layout(available_space);
+        context.set_child_layout(
+            &vertical,
+            Rect::new(
+                Point::new(
+                    available_space.width
+                        .fit_measured(new_control_size.width)
+                        .saturating_sub(scrollbar_layout.width)
+                        .into_signed(),
+                    Px::ZERO,
+                ),
+                scrollbar_layout.into_signed(),
+            ),
+        );
+        let scroll = self.scroll.get_tracking_invalidate(context);
+
+        let start_item = (scroll.y / item_height_upx).floor().get() as usize;
+        let end_item = ((scroll.y + new_control_size.height) / item_height_upx).ceil().get() as usize;
+        let end_item = end_item.min(item_count-1);
+
+        self.visible_range.set(start_item..end_item);
+        
+        let first = self.items.front().map(|t| t.index);
+        let last = self.items.back().map(|t| t.index);
+        let mut closure = |index| {
+            let widget = self.virtual_list.widget_at(index);
+            let mut widget = widget.widget_ref();
+            let mounted = widget.mounted(&mut context.as_event_context());
+            VirtualListItem { index, mounted }
+        };
+        if self.items.is_empty() || first.unwrap() > end_item || last.unwrap() < start_item {
+            self.items.clear();
+            self.items.extend((start_item..=end_item).map(closure));
+        } else {
+            let first = first.expect("List is not empty");
+            let last = last.expect("List is not empty");
+            if first < start_item {
+                while self.items.front().is_some() && self.items.front().expect("Checked is some").index < start_item {
+                    self.items.pop_front();
+                }
+            }
+            if last > end_item {
+                while self.items.back().is_some() && self.items.back().expect("Checked is some").index > end_item {
+                    self.items.pop_back();
+                }
+            }
+            // no extend front :(
```

Comment:
`vec.splice(0..0, items_to_insert)` is essentially extend_front.

Reply:
As I worked on refactoring this today, I realized this was a VecDeque and there indeed is no solution for this.